Add support for returning the result of each job

robmoss opened the pull request:
By passing `results=True` to `parq.run()`, we now collect the return value of each job and return these values to the caller. See `tests/test_results.py` for examples.

This can be convenient for jobs that return small data structures. However, it should probably not be used when running many jobs that each return a very large data structure; in such cases, alternatives such as saving outputs to disk should probably be used.

robmoss committed:
Add support for returning the result of each job

While this can be convenient for jobs that return small data structures,
it should probably not be used when running many jobs that each return a
very large data structure. In such cases, alternatives such as saving
outputs to disk should probably be used.

robmoss committed:
Document the default behaviour of ignoring results

We also now highlight that collecting results may not be an ideal
solution, depending on the number of jobs and the size of the return
values.

thaopletk commented on doc/index.rst:
Could you add a second simple example that returns the return value of each job?

robmoss replied:
Good idea, thanks TK! I've added a commit that does this. What do you think?

thaopletk replied:
It looks good, thank you for adding it!

robmoss replied:
Thanks TK! Now I'm wondering if I should bump the version number from `0.2.3` to `0.2.4` or `0.3.0` when I merge this ...

robmoss committed:
Add an example of obtaining job return values

robmoss committed:
Avoid deadlock when worker processes are killed

By using a timeout when polling for results, and monitoring the exit
code of each worker process, we can avoid deadlock when one or more
worker processes are terminated unexpectedly.

robmoss committed:
Provide a default timeout when polling for results

This means that parq.run() will not deadlock if a worker process is
terminated unexpectedly, unless the user explicitly sets the timeout to
None.

robmoss committed:
Log when sending the worker sentinel

robmoss committed:
Define custom names for worker processes

Each worker is now named "parq-X", where X is the worker number (1..N).

robmoss committed:
Log job exceptions as warnings

By increasing the logging level from INFO to WARNING, we ensure that
these exceptions are visible by default.
